### PR TITLE
fixes cache time handling in apt module

### LINKED ISSUE #1497: apt module: update_cache=yes does nothing
Hello,
I was trying to use the update_cache feature of the apt module to run apt-get update without installing a package. So I've got this basic task:

```
- name: Updating APT cache
   apt: update_cache=yes
```

When the task runs, it says 'ok' but the cache isn't updated at all.
There seem to be the same behavior if update_cache=yes is added with an actual package installation.

I'm using ansible v1.9.1 on CentOS 7 to manage hosts running Ubuntu 14.04LTS.

### PR DESCRIPTION
**Pull request by cloudnull**
This change is in response to issue #1497 where the apt module would not
properly updating the apt cache in some situations and never returned a state
change on cache update when the module was used without or without an item to
be installed or upgraded.

The change simply allows the apt module to update the cache when
`update_cache` option is used without or without a set `cache_valid_time`. If
`cache_valid_time` is set and the on disk mtime for apt cache is ">" the
provided amount of seconds, which now has a default of 0, the apt cache will
be updated. Additionally if no upgrade, package, or deb is installed or
changed but the apt cache is updated the module will return a changed state
which will help users to know that the state of the environment has changed
due to a task operation, even if it was only an apt cahce update.

Signed-off By: Kevin Carter kevin.carter@rackspace.com

fixes #1497

### CODE REVIEW

File: packaging/os/apt.py
Comment (odyssey4me) [inline packaging/os/apt.py]: 'return', perhaps instead of 'retrun'?

Comment (cloudnull) [inline packaging/os/apt.py]: Done.

**Comment by gregdek** [comment]
@mgwilliams Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please give it a +1; at that point, we will add the “priority_reviewed” flag to this PR, which will move it to the top of our queue for potential inclusion.

File: packaging/os/apt.py
Comment (bcoca) [inline packaging/os/apt.py]: this changes the behavior and now setting a cache_valid_time implies updated_cache=True

Comment (cloudnull) [inline packaging/os/apt.py]: @bcoca This does change the behavior by allowing the user to simply set a `cache_valid_time` which would cause the update to happen if the time was older than what was specificed. If you'd think it best I can remove the `or` which would set it back to the old behavior.

**Comment by cloudnull** [comment]
@gregdek & @bcoca I just noticed the _needs_revision_ tag so I've updated the method to remove the `or` in the cache update conditional which should keep the old behavior. Let me know if you find anything else.

**Comment by gregdek** [comment]
Putting back into core_review for @bcoca

**Comment by cloudnull** [comment]
:bump: I'm curious if there's anything here that I need to work on or should update for this patch?

**Comment by bcoca** [comment]
@cloudnull its on my list to review and test, i just have a lot on that list.

**Comment by cloudnull** [comment]
@bcoca no worries. I just got back to the office and was checking in on all the things i have in flight. Thanks for the update! I hope all is well.

**Comment by cloudnull** [comment]
Rebased off of the head of devel because there were merge conflicts.

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by cloudnull** [comment]
I've gone and rebased this change set. Thank you for the initial review. If there's anything else I need to update or you'd like to see please let me know.

**Comment by cloudnull** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

**Comment by cloudnull** [comment]
TYVM @gregdek please let me know if I can help out in any way.

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by cloudnull** [comment]
ready_for_review -- Thank you for the update @gregdek I've rebased the changeset off of the head of the current devel branch. Please let me know if anything else needs to be updated.

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by cloudnull** [comment]
The last push was to update the module to resolve recent changes causing a merge conflict in this PR.

**Comment by cloudnull** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by cloudnull** [comment]
@gregdek bump, and idea when this might be able to be reviewed?

**Comment by gregdek** [comment]
Still up to the core team. Ping @ansible/core.

**Comment by gregdek** [comment]
Also worth perhaps bringing it up at the Core Team Meeting, which is every Tuesday 1900 UTC and Thursday 1500 UTC on #ansible-meeting on freenode.irc.net. 

https://github.com/ansible/community/blob/master/MEETINGS.md

**Comment by cloudnull** [comment]
@gregdek I've added the meeting to my calendar. Thanks again for the update.

**Comment by senorsmile** [comment]
The bug that this PR fixes just hit me.  What's the status on merging?

**Comment by cloudnull** [comment]
@senorsmile ++ I'd love to see this get in for "2.2" though we're waiting on core review.

File: packaging/os/apt.py
Comment (resmo) [inline packaging/os/apt.py]: Why not raise the exception the cache update went wrong?

**Comment by resmo** [comment]
I made a code review and the code looks straight forward and looks more readable to me. I am heading to testing it.

File: packaging/os/apt.py
Comment (cloudnull) [inline packaging/os/apt.py]: Good point, I'll add that back.

**Comment by cloudnull** [comment]
@resmo thanks for the review, i've updated the PR based on your feedback. let me know how it goes.

**Comment by cloudnull** [comment]
The last PR was a rebase off the HEAD of `devel` because CI checks were failing when rebasing this PR.

**Comment by cloudnull** [comment]
fixed syntax problem in py3.4/5 tests:

``` python-traceback
The command "python3.4 -m compileall -fq ." exited with 1.
0.83s$ python3.5 -m compileall -fq .
*** Error compiling './packaging/os/apt.py'...
  File "./packaging/os/apt.py", line 465
    except Exception, e:
                    ^
SyntaxError: invalid syntax
```

**Comment by sivel** [comment]
To properly handle exceptions for both py2.4 through py3, see http://docs.ansible.com/ansible/developing_modules_python3.html#exceptions

**Comment by cloudnull** [comment]
Ah.. I forgot about 2.4 support... :| PR inbound.

**Comment by cloudnull** [comment]
@resmo @sivel -- IDK whats up with **shippable** all but 1 test passed and **travis-ci** seems to be happy. Any pointers?

**Comment by mattclay** [comment]
@cloudnull There was a transient failure on Shippable due to an issue connecting to Docker Hub. I've restarted the build.

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
@cloudnull Hi Kevin, do you have time to rebase and resolve the conflicts? It is on my list for looking at it before release of 2.2.

**Comment by cloudnull** [comment]
@resmo I've rebased this PR. let me know if you need anything else. Thanks again for taking the time to review this.

**Comment by gregdek** [comment]
@cloudnull A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by cloudnull** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @cloudnull for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: packaging/os/apt.py
Comment (resmo) [inline packaging/os/apt.py]: I would prefer the `module.fail_json` as it was.

**Review by resmo — CHANGES_REQUESTED** [review]
Tested on Debian 7 and 8, worked as expected. I would only like to see the use of fail_json if cache update failed

File: packaging/os/apt.py
Comment (cloudnull) [reply packaging/os/apt.py]: Done.

**Review by resmo — APPROVED** [review]

**Comment by resmo** [comment]
LGTM

**Comment by gregdek** [comment]
Thanks again to @cloudnull for this PR. This PR was reviewed by an Ansible member. Marking for inclusion.

However, we would still appreciate if the maintainer(s) @ansible could look over it.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
Thanks!

**Comment by sYnfo** [comment]
@greenbrian I think https://github.com/ansible/ansible-modules-core/blob/845cc1a531610f2db9639393acc7c1663235716e/packaging/os/apt.py#L911 might contain a bug. `retvals` does not necessarily contain `'changed'`, in particular if `install` (https://github.com/ansible/ansible-modules-core/blob/845cc1a531610f2db9639393acc7c1663235716e/packaging/os/apt.py#L439) fails it will certainly not.

**Pull request merged**
